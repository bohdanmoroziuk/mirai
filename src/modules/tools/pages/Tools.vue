<template>
  <div class="tools q-pa-md q-gutter-y-md">
    <div class="flex items-center justify-between q-gutter-x-md">
      <div class="flex items-center q-gutter-x-sm">
        <new-tool-dialog :groups="groups" @add="addTool" />
        <new-group-dialog @add="addGroup" />
      </div>
      <div>
        <tool-search-form />
      </div>
    </div>

    <div
      v-for="(items, group) in tools"
      :key="group"
    >
      <div class="flex items-center no-wrap">
        <q-chip color="primary" text-color="white" :ripple="false">
          {{ group }}
        </q-chip>
        <q-separator style="flex: 1;" horizontal spaced="lg" inset />
        <div class="actions">
          <q-btn
            color="red"
            round
            flat
            icon="delete"
          />
        </div>
      </div>
      <div class="row q-col-gutter-sm">
        <div
          class="col-xs-12 col-sm-6 col-md-4 col-lg-3"
          v-for="tool of items"
          :key="tool.id"
        >
          <tool-card
            :tool="tool"
            @delete="deleteTool"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { onMounted, computed } from 'vue';
import { storeToRefs } from 'pinia';
import { useQuasar } from 'quasar';

import groupBy from 'lodash/fp/groupBy';

import {
  useToolsStore,
  useGroupsStore,
  NewToolDialog,
  NewGroupDialog,
  ToolSearchForm,
  ToolCard,
  ToolId,
  ToolBody,
  GroupName,
  GroupId,
} from 'src/modules/tools';

const $q = useQuasar();

const toolsStore = useToolsStore();

const groupsStore = useGroupsStore();

const { filteredTools } = storeToRefs(toolsStore);

const { groups } = storeToRefs(groupsStore);

const tools = computed(() => groupBy('group.name', filteredTools.value));

const getTools = async () => {
  await toolsStore.getTools();
};

const addTool = async (body: ToolBody) => {
  try {
    await toolsStore.addTool(body);

    $q.notify({
      type: 'positive',
      message: 'Added a new tool',
    });
  } catch (error) {
    $q.notify({
      type: 'negative',
      message: (error as Error).message,
    });
  }
};

const deleteTool = async (id: ToolId) => {
  try {
    await toolsStore.deleteTool(id);

    $q.notify({
      type: 'positive',
      message: 'Tool successfully deleted',
    });
  } catch (error) {
    $q.notify({
      type: 'negative',
      message: (error as Error).message,
    });
  }
};

const getGroups = async () => {
  await groupsStore.getGroups();
};

const addGroup = async (name: GroupName) => {
  try {
    await groupsStore.addGroup(name);

    $q.notify({
      type: 'positive',
      message: 'Group successfully deleted',
    });
  } catch (error) {
    $q.notify({
      type: 'negative',
      message: (error as Error).message,
    });
  }
};

const deleteGroup = async (id: GroupId) => {
  try {
    await groupsStore.deleteGroup(id);

    $q.notify({
      type: 'positive',
      message: 'Group successfully deleted',
    });
  } catch (error) {
    $q.notify({
      type: 'negative',
      message: (error as Error).message,
    });
  }
};

const getToolsAndGroups = async () => {
  await getTools();
  await getGroups();
};

onMounted(getToolsAndGroups);
</script>
