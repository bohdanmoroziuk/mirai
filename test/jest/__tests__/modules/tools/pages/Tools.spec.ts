import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-jest';
import {
  beforeEach,
  describe,
  expect,
  jest,
  it,
} from '@jest/globals';
import { mount, flushPromises } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

import Tools from 'src/modules/tools/pages/Tools.vue';

installQuasarPlugin();

const mockTools = [
  {
    id: '1',
    name: 'React',
    url: 'https://uk.reactjs.org/',
  },
  {
    id: '2',
    name: 'Vue',
    url: 'https://vuejs.org/',
  },
];

jest.mock('src/modules/tools/services/tools.ts', () => ({
  toolsService: {
    getTools: jest.fn().mockImplementation(() => Promise.resolve(mockTools)),
    addTool: jest.fn().mockImplementation(() => Promise.resolve()),
    deleteTool: jest.fn().mockImplementation(() => Promise.resolve()),
  },
}));

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Tools', () => {
  it('renders successfully', async () => {
    const wrapper = mount(Tools, {
      global: {
        plugins: [
          createTestingPinia({
            stubActions: false,
          }),
        ],
      },
    });

    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays a list of tools', async () => {
    const wrapper = mount(Tools, {
      global: {
        plugins: [
          createTestingPinia({
            stubActions: false,
          }),
        ],
      },
    });

    await flushPromises();

    expect(wrapper.text()).toContain(mockTools[0].name);
    expect(wrapper.text()).toContain(mockTools[1].name);
  });
});
