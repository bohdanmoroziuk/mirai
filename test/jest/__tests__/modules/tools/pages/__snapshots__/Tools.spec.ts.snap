// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tools renders successfully 1`] = `
<div class="tools q-pa-md q-gutter-y-md">
  <div class="flex items-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-primary text-white q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon on-left notranslate material-icons" aria-hidden="true" role="img">add</i><span class="block">Add tool</span></span></button>
    <!---->
  </div>
  <div class="tool-list">
    <div class="row q-col-gutter-sm">
      <div class="col-xs-12 col-sm-6 col-md-4 col-lg-3">
        <div class="q-card q-card--bordered q-card--flat no-shadow cursor-pointer">
          <div class="q-card__section q-card__section--vert q-px-xs q-py-md">
            <div class="flex justify-center"><img class="image" src="https://cdn.quasar.dev/logo-v2/svg/logo.svg" alt="React"></div>
          </div>
          <div class="q-card__section q-card__section--vert q-pa-xs q-pl-md">
            <div class="flex items-center no-wrap">
              <div class="text-caption text-primary">React</div>
              <div class="q-space"></div><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-red q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">delete</i></span></button>
            </div>
          </div>
        </div>
      </div>
      <div class="col-xs-12 col-sm-6 col-md-4 col-lg-3">
        <div class="q-card q-card--bordered q-card--flat no-shadow cursor-pointer">
          <div class="q-card__section q-card__section--vert q-px-xs q-py-md">
            <div class="flex justify-center"><img class="image" src="https://cdn.quasar.dev/logo-v2/svg/logo.svg" alt="Vue"></div>
          </div>
          <div class="q-card__section q-card__section--vert q-pa-xs q-pl-md">
            <div class="flex items-center no-wrap">
              <div class="text-caption text-primary">Vue</div>
              <div class="q-space"></div><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-red q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">delete</i></span></button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
